gpmp2/kinematics/WorkspacePoseVelocityPrior.h

```diff
@@ -0,0 +1,125 @@
+/**
+ *  @file   WorkspacePoseVelocityPrior.h
+ *  @brief  Prior defined on the workspace pose and velocity of final joint
+ *(end-effector) of a robot given its state in configuration space
+ *  @author Matthew King-Smith
+ *  @date   May 16, 2023
+ **/
+
+#pragma once
+
+#include <gpmp2/kinematics/Arm.h>
+#include <gtsam/base/Matrix.h>
+#include <gtsam/base/Vector.h>
+#include <gtsam/geometry/Pose3.h>
+#include <gtsam/nonlinear/NonlinearFactor.h>
+
+#include <iostream>
+#include <vector>
+
+namespace gpmp2 {
+
+/**
+ * binary factor Gaussian prior defined on the workspace pose and velocities
+ */
+class WorkspacePoseVelocityPrior
+    : public gtsam::NoiseModelFactor2<gtsam::Vector, gtsam::Vector> {
+ private:
+  // typedefs
+  typedef WorkspacePoseVelocityPrior This;
+  typedef gtsam::NoiseModelFactor2<gtsam::Vector, gtsam::Vector> Base;
+
+  // arm
+  Arm arm_;
+
+  // desired workspace pose
+  gtsam::Pose3 des_pose_;
+
+  // desired velocity (linear and angular)
+  gtsam::Vector6 des_vel_;
+
+ public:
+  /// shorthand for a smart pointer to a factor
+  typedef std::shared_ptr<This> shared_ptr;
+
+  /// Default constructor
+  WorkspacePoseVelocityPrior() {}
+
+  /**
+   * Constructor
+   * @param cost_model cost function covariance
+   * @param des_vel desired velocity (linear and angular)
+   */
+  WorkspacePoseVelocityPrior(gtsam::Key poseKey, gtsam::Key velKey,
+                             const gtsam::SharedNoiseModel& cost_model,
+                             const Arm& arm, const gtsam::Pose3& des_pose,
+                             const gtsam::Vector6& des_vel)
+      : Base(cost_model, poseKey, velKey),
+        arm_(arm),
+        des_pose_(des_pose),
+        des_vel_(des_vel) {}
+
+  ~WorkspacePoseVelocityPrior() {}
+
+  /// error function
+  gtsam::Vector evaluateError(
+      const gtsam::Vector& joint_conf, const gtsam::Vector& joint_rates,
+      gtsam::OptionalMatrixType H1 = nullptr,
+      gtsam::OptionalMatrixType H2 = nullptr) const override {
+    using namespace gtsam;
+
+    // fk
+    std::vector<Pose3> joint_pose;
+    std::vector<Vector6> joint_vel;
+    std::vector<Matrix> J_jpx_jp, J_jvx_jp, J_jvx_jv;
+    arm_.forwardKinematics(joint_conf, joint_rates, joint_pose, &joint_vel,
+                           &J_jpx_jp, &J_jvx_jp, &J_jvx_jv);
+
+    Matrix66 H_ep;
+    Vector pose_error = des_pose_.logmap(joint_pose[arm_.dof() - 1], {}, H_ep);
+    Vector vel_error = des_vel_ - joint_vel[arm_.dof() - 1];
+    if (H1) {
+      // Jacobian for the joint positions
+      *H1 = (Matrix(12, arm_.dof()) << H_ep * J_jpx_jp[arm_.dof() - 1],//
+                              - J_jvx_jp[arm_.dof() - 1]).finished();
```

Comment:
Some weird formatting issue is happening here. Can you please reformat the file?